### ISSUE TITLE
Low number of identified peptides using the RESET algorithm

### ISSUE THREAD

**Issue by mhamaneh** [percolator/percolator#382]
Hello,
I am using the latest version (3.07.01) of Percolator and trying to run the RESET algorithm using the following command:

percolator -Y --reset-algorithm --num-threads 32 --subset-max-train 500000 --post-processing-tdc --testFDR 0.01 --trainFDR 0.01 --results-peptides peptides --results-psms psms --decoy-results-peptides decoy_peptides --decoy-results-psms decoy_psms pin.tab

where pin.tab is a PIN file, with user-defined features, containing both target and decoy peptides. The problem is that the number of identified peptides is significantly (more than 30%) lower using the RESET algorithm in comparison with that obtained using the original algorithm (same command without --reset-algorithm). Based on my understanding of the RESET algorithm publication, I expected the number of identified peptides to be comparable in both cases. Also, if, instead of running percolator+RESET, I rank the peptides using just one of the features I get more peptides at the 0.01 level than when I run Percolator with RESET. This is does not make any sense to me. Am I doing something wrong? 

Thank you very much in advance for your help.

**Comment by freejstone** [comment]
Hi there, are you able to share the log information and the pin file so that I can further troubleshoot? I have a separate implementation of percolator-reset that I can use to check/compare what might be going on.

Thanks!

**Comment by mhamaneh** [comment]
Hi,
Thank you very much for the prompt reply. Unfortunately, the PIN file is too large to share. I wanted to share a small portion of it but, I could not reproduce the problem using this small file, i.e. the number of identified peptides were comparable whether or not the RESET algorithm was used. I ran Percolator and Percolator+RESET for some other files and noticed that a significant (more than 10%) difference between the numbers of identified peptides is observed only when the PIN file is large (more than 1000000 PSMs), and that the relative difference is larger for larger files. I have pasted the log for one PIN files below (with close to 3000000 PSMs). Note that the number of identified peptides in this case is 56253. However, without RESET I get more than 72000 peptides. 

Percolator version 3.07.1, Build Date Jul  2 2024 12:20:21
Copyright (c) 2006-9 University of Washington. All rights reserved.
Written by Lukas KÃ¤ll (lukall@u.washington.edu) in the
Department of Genome Sciences at the University of Washington.
Issued command:
percolator -Y --reset-algorithm --num-threads 32 --subset-max-train 500000 --post-processing-tdc --testFDR 0.01 --trainFDR 0.01 --results-peptides peptides --results-psms psms --decoy-results-peptides decoy_peptides --decoy-results-psms decoy_psms new_pin.tab
Started Thu Jul 18 10:08:08 2024
 on biowulf.nih.gov
Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
Finding protein decoy prefix for new_pin.tab
Using protein decoy prefix ""
Reading tab-delimited input from datafile new_pin.tab
Features:
Mass missedCleavages sequence_length KR Charge1 Charge2 Charge3 Charge4 Charge5 Charge6 cosine count_predicted count_predicted_b count_predicted_y count_observed count_observed_b count_observed_y count_observed_and_predicted count_observed_and_predicted_b count_observed_and_predicted_y count_not_observed_and_not_predicted count_not_observed_and_not_predicted_b count_not_observed_and_not_predicted_y count_observed_but_not_predicted count_observed_but_not_predicted_b count_observed_but_not_predicted_y count_not_observed_but_predicted count_not_observed_but_predicted_b count_not_observed_but_predicted_y fraction_predicted fraction_predicted_b fraction_predicted_y fraction_observed fraction_observed_b fraction_observed_y fraction_observed_and_predicted fraction_observed_and_predicted_b fraction_observed_and_predicted_y fraction_not_observed_and_not_predicted fraction_not_observed_and_not_predicted_b fraction_not_observed_and_not_predicted_y fraction_observed_but_not_predicted fraction_observed_but_not_predicted_b fraction_observed_but_not_predicted_y fraction_not_observed_but_predicted fraction_not_observed_but_predicted_b fraction_not_observed_but_predicted_y fraction_observed_and_predicted_vs_predicted fraction_observed_and_predicted_b_vs_predicted_b fraction_observed_and_predicted_y_vs_predicted_y fraction_not_observed_and_not_predicted_vs_predicted fraction_not_observed_and_not_predicted_b_vs_predicted_b fraction_not_observed_and_not_predicted_y_vs_predicted_y fraction_observed_but_not_predicted_vs_predicted fraction_observed_but_not_predicted_b_vs_predicted_b fraction_observed_but_not_predicted_y_vs_predicted_y fraction_not_observed_but_predicted_vs_predicted fraction_not_observed_but_predicted_b_vs_predicted fraction_not_observed_but_predicted_y_vs_predicted 
Processing line 1000000
Processing line 2000000
Found 2934990 PSMs
Separate target and decoy search inputs detected, using target-decoy competition on Percolator scores.
Train/test set contains 353737 positives and 146263 negatives, size ratio=2.4185 and pi0=1
Running the Percolator-RESET algorithm.
Selecting best separating single variable.
No weights file given, using default direction
Selected feature 20 as initial direction. Could separate 185027 training set positives with q<0.01 in that direction.
Starting reset: psmsOnly
Selected the best scoring PSM for each of the 248450 peptides from a dataset of 500000 PSMs.
Splitting into train/test
Setting up SVM training for a size of 248450 peptides.
Training set prepared. Starting SVM Training.
GridSearch found the optimal hyperParameters for SVM training, C+=100, and C-/C+=3.
Training with a set of size 100978 preptides, where of 52402 are target, and 48576 are decoys.
Training with a set of size 101683 preptides, where of 53107 are target, and 48576 are decoys.
Training with a set of size 102035 preptides, where of 53459 are target, and 48576 are decoys.
Training with a set of size 102230 preptides, where of 53654 are target, and 48576 are decoys.
Training with a set of size 102351 preptides, where of 53775 are target, and 48576 are decoys.
Training Done!
Final feature weights
Mass    -0.0323436
missedCleavages 0.889996
sequence_length 1.64104
KR      -1.27546
Charge1 -0.241573
Charge2 -0.0036984
Charge3 0.0790102
Charge4 -0.0874623
Charge5 0.0886571
Charge6 -0.0196453
cosine  1.23625
count_predicted 0.0495139
count_predicted_b       -0.286178
count_predicted_y       0.288959
count_observed  0.68139
count_observed_b        0.200365
count_observed_y        0.831636
count_observed_and_predicted    1.08457
count_observed_and_predicted_b  0.473306
count_observed_and_predicted_y  1.20283
count_not_observed_and_not_predicted    -0.141964
count_not_observed_and_not_predicted_b  -0.094023
count_not_observed_and_not_predicted_y  -0.192791
count_observed_but_not_predicted        -0.669731
count_observed_but_not_predicted_b      -0.24073
count_observed_but_not_predicted_y      -0.882553
count_not_observed_but_predicted        -0.669272
count_not_observed_but_predicted_b      -0.51154
count_not_observed_but_predicted_y      -0.623852
fraction_predicted      -0.199469
fraction_predicted_b    0.0173042
fraction_predicted_y    -0.313223
fraction_observed       -0.157169
fraction_observed_b     -0.154446
fraction_observed_y     -0.138279
fraction_observed_and_predicted -0.180135
fraction_observed_and_predicted_b       -0.212171
fraction_observed_and_predicted_y       -0.147823
fraction_not_observed_and_not_predicted 0.190416
fraction_not_observed_and_not_predicted_b       -0.0204601
fraction_not_observed_and_not_predicted_y       0.305861
fraction_observed_but_not_predicted     -0.00652703
fraction_observed_but_not_predicted_b   0.0113937
fraction_observed_but_not_predicted_y   -0.0217749
fraction_not_observed_but_predicted     -0.0127645
fraction_not_observed_but_predicted_b   0.134282
fraction_not_observed_but_predicted_y   -0.109995
fraction_observed_and_predicted_vs_predicted    0.58986
fraction_observed_and_predicted_b_vs_predicted_b        -0.317894
fraction_observed_and_predicted_y_vs_predicted_y        -0.193514
fraction_not_observed_and_not_predicted_vs_predicted    -1.79326
fraction_not_observed_and_not_predicted_b_vs_predicted_b        0.373042
fraction_not_observed_and_not_predicted_y_vs_predicted_y        0.412952
fraction_observed_but_not_predicted_vs_predicted        0.370143
fraction_observed_but_not_predicted_b_vs_predicted_b    -0.00698986
fraction_observed_but_not_predicted_y_vs_predicted_y    0.448182
fraction_not_observed_but_predicted_vs_predicted        -0.58986
fraction_not_observed_but_predicted_b_vs_predicted      0.246851
fraction_not_observed_but_predicted_y_vs_predicted      0.419746
Intercept       -1.05137
Test set evaluation finds 56253 peptides under a FDR of 0.01
Processing took 182.4777 cpu seconds or 44 seconds wall clock time.

**Comment by freejstone** [comment]
Thanks for the log file. Could you share the smaller files still? Or is it possible to share the large data set on a separate platform? It will be difficult to diagnose the problem otherwise.

It's very strange that it only happens when the file is large. I have previously verified on a large data set that the number of discoveries are very comparable (discoveries on the order of 100K peptides). I have a hunch that it might be to do with the user specified features.

**Comment by mhamaneh** [comment]
[pin.tab.gz](https://github.com/user-attachments/files/16312288/pin.tab.gz)
I have attached the first 100000 lines of the file.

**Comment by freejstone** [comment]
If the first 100K lines are representative of the rest of the file I am getting very similar results for percolator and percolator-RESET as you do.

Here are some possibilities:

- @percolator is it possible that the newly added RESET throws out rows from being discovered after it reaches `--subset-max-train 500000` ? It seems to just be a problem on large data sets and this would account for this unusual behaviour. As a sanity check @mhamaneh , because I can't access your large file, you can maybe filter your `.pin` file for the best scoring PSM for each peptide, and hopefully this reduces the number of rows to below 500K. It appears from the log file that feature 20 is the best scoring one `count_observed_and_predicted_y` for this job. Then we can see how percolator-RESET works on this filtered pin file (and maybe it might be worth seeing how percolator does on this filtered pin file as well).
- You may want to try upping the number of iterations, `--maxiter`. In the paper, 5 iterations seemed to suffice but given that you are using much more data with many more features, it could be worth bumping this up.

Thanks!

**Comment by mhamaneh** [comment]
Thanks again for the prompt reply. A couple of questions/comments:
(1) Why not just increase `--subset-max-train` instead of filtering the PIN file? I increased this parameter first to 1000000, and then in increments of 1000000, to 5000000. The number of peptides identified by percolator-RESET increased until `--subset-max-train` reached 3000000, which is roughly the number of PSMs in the PIN file. Increasing the parameter more than that did not have a significant effect. Also, I observed almost no change when I used percolator with larger values for `--subset-max-train`. In fact, I get slightly (around 7%) more peptides from percolator-RESET than from percolator when `--subset-max-train` is set to 3000000. Does this make sense? Should I just increase `--subset-max-train` to be at least equal to the number lines in the file? Do I still get accurate FDRs?
(2) Increasing the number of iterations had almost no effect.

**Comment by freejstone** [comment]
(1) You're right, certainly increasing `--subset-max-train` would work. It definitely sounds like a bug to me, that is, the reported discoveries are only on the trained peptides set by `--subset-max-train`. Having more discoveries isn't necessarily an indicator that FDR is not controlled. percolator-RESET and percolator do things quite a bit different. I have found an increase of 2-4% in some smaller datasets, though 7% is a little more. However, I have not investigated all types of search engines, or the user-defined features you use.

RESET has theoretical FDR control under the assumptions that essentially your features are not able to distinguish between the incorrect target peptides and the decoys. The problem with using user-define features is that they have not been tested against this assumption. So if you can justify or have no reason to believe that your features are able to distinguish between incorrect target peptides and decoys, then to me it seems valid.

(2) No worries.

@percolator Hey Lukas, just tagging you here. I believe there seems to be quite a substantial bug.

**Comment by mhamaneh** [comment]
Thanks again for the explanation. Just one more question: does setting `--subset-max-train` equal to the total number PSMs mean that percolator-RESET uses all PSMs for training? If yes, isn't that a problem (since a subset should be used for training)?

**Comment by freejstone** [comment]
My understanding is that `--subset-max-train` just controls the amount of PSM objects that enter the percolator algorithm for computational efficiency. In the context of percolator RESET, it is fine for all PSM objects to enter the percolator RESET algorithm, so setting `--subset-max-train` to be equal to the total number of PSMs has no problems. Internally there is a data-splitting procedure (the RESET component of percolator-RESET) that ensures FDR control.

**Comment by mhamaneh** [comment]
Great! thanks. Just one more comment: The log of percolator-RESET reports the total number of peptides (target AND decoy) with FDR<0.01, whereas percolator reports only the number of identified target peptides, which makes more sense to me. I am not sure if this is intentional, so I thought I should let you know.

**Comment by freejstone** [comment]
Yeah that doesn't sound right. I will forward this onto @percolator, thanks!

**Comment by percolator** [comment]
Yes, this is a bug. We should disable the subset training options when using percolator-reset, However, I don't fully see the point of the combination. The risk of overfitting practically disappears for larger datasets.

**Comment by mhamaneh** [comment]
Hello again,

I have another question regarding Percolator-RESET. When I run it using this command:
`percolator -Y --reset-algorithm --results-peptides peptides --results-psms psms --decoy-results-peptides decoy_peptides --decoy-results-psms decoy_psms pin.tab `
one of the lines in the output states "Starting reset: psmsOnly" and indeed the resulting "peptides" file is empty. But when I look at the "psms" file I see only non-redundant peptides and q-values seem to have been calculated for peptides rather than psms. So, I am confused are these q-values for peptides or psms? If they are associated with psms, why are there no non-redundant peptides? Also, if the q-values are reported at the psms levels only, how do I get the q-values at the peptide level?

Thank you very much in advance for your help.
